File: apps/reading/src/api/midnightOilSourceAttachQualityCompose.ts
/**
 * Midnight Oil unattended + source attach quality pack (pure).
 *
 * Operator vision: set time + goals + price ceiling for unattended deep
 * research ("midnight oil"), while attaching knowledge-dense HTML sources
 * (arxiv/substack) under quality/budget honesty — highest-quality DR product
 * without needing the operator present in the workstation.
 *
 * live_execution_authorized always false.
 * remote_fetched always false.
 * pdf_view_authorized always false.
 * live_dispatched always false.
 * store_mutated always false.
 */

import {
  composeMidnightOilUnattendedPackage,
  type MidnightOilUnattendedPackageCompose,
  type MoGoalEntry,
} from "./midnightOilUnattendedPackageCompose";
import {
  composeSourcePublicationDrAttachQuality,
  type SourcePublicationDrAttachQualityCompose,
} from "./sourcePublicationDrAttachQualityCompose";
import type { HtmlNativeSourceRef } from "./htmlNativeSourceAttachCompose";
import type { CitationRecord } from "./deepResearchSourceCitationPack";
import type { PublicationFamily } from "./sourcePublicationRegistry";

export type { MoGoalEntry };

export interface MidnightOilSourceAttachQualityInput {
  operator_id: string;
  work_minutes: number;
  goals: MoGoalEntry[];
  usd_per_hour?: number | null;
  approved_ceiling_usd?: number | null;
  operator_ack: boolean;
  unattended_ack: boolean;
  spend_consent: boolean;
  brief_dispatch_ready?: boolean;
  /** Research session / parent for source attach. */
  session_id: string;
  parent_asset_id: string;
  requested_families: PublicationFamily[];
  sources: HtmlNativeSourceRef[];
  citations?: CitationRecord[] | null;
  derive_citations_from_sources?: boolean;
  quality_overall: number | null;
  quality_floor?: number;
  would_exceed: boolean | null;
  operator_override?: boolean;
  /**
   * When true (default), require MO unattended_package_ready AND source pack_ready.
   */
  require_both?: boolean;
}

export interface MidnightOilSourceAttachQualityCompose {
  operator_id: string;
  session_id: string;
  parent_asset_id: string;
  mo_unattended: MidnightOilUnattendedPackageCompose;
  source_quality: SourcePublicationDrAttachQualityCompose;
  /**
   * True when mo ready and source pack ready (or either if require_both=false)
   * and operator_ack + unattended_ack + spend_consent path still pure.
   */
  pack_ready: boolean;
  live_execution_authorized: false;
  remote_fetched: false;
  pdf_view_authorized: false;
  live_dispatched: false;
  store_mutated: false;
  notes: string[];
  authority: "midnight_oil_source_attach_quality_compose_advisory";
}

function requireNonEmpty(value: unknown, name: string): string {
  if (typeof value !== "string" || !value.trim()) {
    throw new Error(`${name} must be a non-empty string`);
  }
  return value.trim();
}

/**
 * Compose MO unattended package + arxiv/substack source quality attach.
 * Never launches workers; never scrapes; never charges.
 */
export function composeMidnightOilSourceAttachQuality(
  input: MidnightOilSourceAttachQualityInput,
): MidnightOilSourceAttachQualityCompose {
  if (!input || typeof input !== "object") {
    throw new Error("input must be an object");
  }
  if (typeof input.operator_ack !== "boolean") {
    throw new Error("operator_ack must be an explicit boolean");
  }
  if (typeof input.unattended_ack !== "boolean") {
    throw new Error("unattended_ack must be an explicit boolean");
  }
  if (typeof input.spend_consent !== "boolean") {
    throw new Error("spend_consent must be an explicit boolean");
  }
  const operator_id = requireNonEmpty(input.operator_id, "operator_id");
  const session_id = requireNonEmpty(input.session_id, "session_id");
  const parent_asset_id = requireNonEmpty(
    input.parent_asset_id,
    "parent_asset_id",
  );

  const require_both =
    input.require_both === undefined ? true : input.require_both;
  if (typeof require_both !== "boolean") {
    throw new Error("require_both must be boolean when set");
  }

  const notes: string[] = [
    "live_execution_authorized=false — midnight oil never launches workers",
    "remote_fetched=false — no live arxiv/substack scrape",
    "pdf_view_authorized=false — HTML-native sources only",
    "live_dispatched=false · store_mutated=false",
  ];

  const mo_unattended = composeMidnightOilUnattendedPackage({
    operator_id,
    work_minutes: input.work_minutes,
    goals: input.goals,
    usd_per_hour: input.usd_per_hour,
    approved_ceiling_usd: input.approved_ceiling_usd,
    operator_ack: input.operator_ack,
    unattended_ack: input.unattended_ack,
    spend_consent: input.spend_consent,
    brief_dispatch_ready: input.brief_dispatch_ready,
  });
  notes.push(...mo_unattended.notes.map((n) => `[mo] ${n}`));

  const source_quality = composeSourcePublicationDrAttachQuality({
    session_id,
    parent_asset_id,
    requested_families: input.requested_families,
    sources: input.sources,
    citations: input.citations,
    derive_citations_from_sources: input.derive_citations_from_sources,
    quality_overall: input.quality_overall,
    quality_floor: input.quality_floor,
    would_exceed: input.would_exceed,
    operator_override: input.operator_override,
    operator_ack: input.operator_ack,
  });
  notes.push(...source_quality.notes.map((n) => `[source_quality] ${n}`));

  let pack_ready = false;
  if (require_both) {
    pack_ready =
      mo_unattended.unattended_package_ready === true &&
      source_quality.pack_ready === true &&
      input.operator_ack === true;
  } else {
    pack_ready =
      input.operator_ack === true &&
      (mo_unattended.unattended_package_ready === true ||
        source_quality.pack_ready === true);
  }

  if (pack_ready) {
    notes.push(
      "pack_ready=true — MO unattended + source quality ready; still pure",
    );
  } else {
    notes.push(
      "pack_ready=false — MO package, source quality, or operator_ack gate open",
    );
  }

  if (
    mo_unattended.live_execution_authorized !== false ||
    source_quality.remote_fetched !== false ||
    source_quality.pdf_view_authorized !== false ||
    source_quality.live_dispatch_authorized !== false ||
    source_quality.store_mutated !== false
  ) {
    throw new Error("invariant: honesty flags must remain false");
  }

  notes.push("live_execution_authorized=false");
  notes.push("remote_fetched=false");
  notes.push("pdf_view_authorized=false");
  notes.push("live_dispatched=false");
  notes.push("store_mutated=false");

  return {
    operator_id,
    session_id,
    parent_asset_id,
    mo_unattended,
    source_quality,
    pack_ready,
    live_execution_authorized: false,
    remote_fetched: false,
    pdf_view_authorized: false,
    live_dispatched: false,
    store_mutated: false,
    notes,
    authority: "midnight_oil_source_attach_quality_compose_advisory",
  };
}

export function formatMidnightOilSourceAttachQualitySummary(
  c: MidnightOilSourceAttachQualityCompose,
): string {
  return (
    `pack_ready=${c.pack_ready} · ` +
    `mo_ready=${c.mo_unattended.unattended_package_ready} · ` +
    `source_ready=${c.source_quality.pack_ready} · ` +
    `sources=${c.source_quality.attach.source_count} · ` +
    `live_execution_authorized=false · remote_fetched=false · pdf_view_authorized=false`
  );
}

File: substrate/midnight_oil_source_attach_quality_compose.py
"""Midnight Oil unattended + source attach quality pack (pure).

live_execution_authorized / remote_fetched / pdf_view_authorized /
live_dispatched / store_mutated always False.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any

from substrate.midnight_oil_unattended_package_compose import (
    MidnightOilUnattendedPackageCompose,
    MidnightOilUnattendedPackageComposeError,
    compose_midnight_oil_unattended_package,
)
from substrate.source_publication_dr_attach_quality_compose import (
    SourcePublicationDrAttachQualityCompose,
    SourcePublicationDrAttachQualityComposeError,
    compose_source_publication_dr_attach_quality,
)


class MidnightOilSourceAttachQualityComposeError(ValueError):
    """Fail-closed validation for MO + source attach quality pack."""


@dataclass(frozen=True)
class MidnightOilSourceAttachQualityCompose:
    operator_id: str
    session_id: str
    parent_asset_id: str
    mo_unattended: MidnightOilUnattendedPackageCompose
    source_quality: SourcePublicationDrAttachQualityCompose
    pack_ready: bool
    live_execution_authorized: bool
    remote_fetched: bool
    pdf_view_authorized: bool
    live_dispatched: bool
    store_mutated: bool
    notes: tuple[str, ...]
    authority: str

    def to_dict(self) -> dict[str, Any]:
        return {
            "operator_id": self.operator_id,
            "session_id": self.session_id,
            "parent_asset_id": self.parent_asset_id,
            "mo_unattended": self.mo_unattended.to_dict(),
            "source_quality": self.source_quality.to_dict(),
            "pack_ready": self.pack_ready,
            "live_execution_authorized": False,
            "remote_fetched": False,
            "pdf_view_authorized": False,
            "live_dispatched": False,
            "store_mutated": False,
            "notes": list(self.notes),
            "authority": "midnight_oil_source_attach_quality_compose_advisory",
        }


def _require_nonempty(value: object, *, field: str) -> str:
    if not isinstance(value, str) or not value.strip():
        raise MidnightOilSourceAttachQualityComposeError(
            f"{field} must be a non-empty string"
        )
    return value.strip()


def compose_midnight_oil_source_attach_quality(
    *,
    operator_id: object,
    work_minutes: object,
    goals: object,
    operator_ack: object,
    unattended_ack: object,
    spend_consent: object,
    session_id: object,
    parent_asset_id: object,
    requested_families: object,
    sources: object,
    quality_overall: object,
    would_exceed: object,
    usd_per_hour: object | None = None,
    approved_ceiling_usd: object | None = None,
    brief_dispatch_ready: object | None = None,
    citations: object | None = None,
    derive_citations_from_sources: object | None = None,
    quality_floor: object | None = None,
    operator_override: object | None = None,
    require_both: object | None = None,
) -> MidnightOilSourceAttachQualityCompose:
    """MO unattended + HTML source quality. Never launches/scrapes."""
    if not isinstance(operator_ack, bool):
        raise MidnightOilSourceAttachQualityComposeError(
            "operator_ack must be an explicit boolean"
        )
    if not isinstance(unattended_ack, bool):
        raise MidnightOilSourceAttachQualityComposeError(
            "unattended_ack must be an explicit boolean"
        )
    if not isinstance(spend_consent, bool):
        raise MidnightOilSourceAttachQualityComposeError(
            "spend_consent must be an explicit boolean"
        )
    op = _require_nonempty(operator_id, field="operator_id")
    session = _require_nonempty(session_id, field="session_id")
    parent = _require_nonempty(parent_asset_id, field="parent_asset_id")

    require = True if require_both is None else require_both
    if not isinstance(require, bool):
        raise MidnightOilSourceAttachQualityComposeError(
            "require_both must be boolean when set"
        )

    notes: list[str] = [
        "live_execution_authorized=false — midnight oil never launches workers",
        "remote_fetched=false — no live arxiv/substack scrape",
        "pdf_view_authorized=false — HTML-native sources only",
        "live_dispatched=false · store_mutated=false",
    ]

    try:
        mo_unattended = compose_midnight_oil_unattended_package(
            operator_id=op,
            work_minutes=work_minutes,
            goals=goals,
            operator_ack=operator_ack,
            unattended_ack=unattended_ack,
            spend_consent=spend_consent,
            usd_per_hour=usd_per_hour,
            approved_ceiling_usd=approved_ceiling_usd,
            brief_dispatch_ready=brief_dispatch_ready,
        )
    except MidnightOilUnattendedPackageComposeError as e:
        raise MidnightOilSourceAttachQualityComposeError(str(e)) from e
    notes.extend(f"[mo] {n}" for n in mo_unattended.notes)

    try:
        source_quality = compose_source_publication_dr_attach_quality(
            session_id=session,
            parent_asset_id=parent,
            requested_families=requested_families,
            sources=sources,
            quality_overall=quality_overall,
            would_exceed=would_exceed,
            operator_ack=operator_ack,
            citations=citations,
            derive_citations_from_sources=derive_citations_from_sources,
            quality_floor=quality_floor,
            operator_override=operator_override,
        )
    except SourcePublicationDrAttachQualityComposeError as e:
        raise MidnightOilSourceAttachQualityComposeError(str(e)) from e
    notes.extend(f"[source_quality] {n}" for n in source_quality.notes)

    if require:
        pack_ready = (
            mo_unattended.unattended_package_ready is True
            and source_quality.pack_ready is True
            and operator_ack is True
        )
    else:
        pack_ready = operator_ack is True and (
            mo_unattended.unattended_package_ready is True
            or source_quality.pack_ready is True
        )

    if pack_ready:
        notes.append(
            "pack_ready=true — MO unattended + source quality ready; still pure"
        )
    else:
        notes.append(
            "pack_ready=false — MO package, source quality, or operator_ack "
            "gate open"
        )

    if (
        mo_unattended.live_execution_authorized is not False
        or source_quality.remote_fetched is not False
        or source_quality.pdf_view_authorized is not False
        or source_quality.live_dispatch_authorized is not False
        or source_quality.store_mutated is not False
    ):
        raise MidnightOilSourceAttachQualityComposeError(
            "invariant: honesty flags must remain false"
        )

    notes.extend(
        (
            "live_execution_authorized=false",
            "remote_fetched=false",
            "pdf_view_authorized=false",
            "live_dispatched=false",
            "store_mutated=false",
        )
    )

    return MidnightOilSourceAttachQualityCompose(
        operator_id=op,
        session_id=session,
        parent_asset_id=parent,
        mo_unattended=mo_unattended,
        source_quality=source_quality,
        pack_ready=pack_ready,
        live_execution_authorized=False,
        remote_fetched=False,
        pdf_view_authorized=False,
        live_dispatched=False,
        store_mutated=False,
        notes=tuple(notes),
        authority="midnight_oil_source_attach_quality_compose_advisory",
    )


def format_midnight_oil_source_attach_quality_summary(
    c: MidnightOilSourceAttachQualityCompose,
) -> str:
    return (
        f"pack_ready={c.pack_ready} · "
        f"mo_ready={c.mo_unattended.unattended_package_ready} · "
        f"source_ready={c.source_quality.pack_ready} · "
        f"sources={c.source_quality.attach.source_count} · "
        f"live_execution_authorized=false · remote_fetched=false · "
        f"pdf_view_authorized=false"
    )


__all__ = [
    "MidnightOilSourceAttachQualityCompose",
    "MidnightOilSourceAttachQualityComposeError",
    "compose_midnight_oil_source_attach_quality",
    "format_midnight_oil_source_attach_quality_summary",
]
